gnovm/pkg/gnolang/stacktrace.go

```diff
@@ -0,0 +1,81 @@
+package gnolang
+
+import (
+	"fmt"
+	"strings"
+)
+
+const maxStacktraceSize = 128
+
+type Execution struct {
+	Stmt  Stmt
+	Frame *Frame
+}
+type Stacktrace struct {
+	Executions      []Execution
+	NumFramesElided int
+}
+
+func (s Stacktrace) String() string {
+	var builder strings.Builder
+
+	for i, e := range s.Executions {
+		if s.NumFramesElided > 0 && i == maxStacktraceSize/2 {
+			builder.WriteString(fmt.Sprintf("...%d frame(s) elided...\n", s.NumFramesElided))
+		}
+
+		switch {
+		case e.Stmt == nil:
+			builder.WriteString(fmt.Sprintf("%s()\n", e.Frame.Func))
+			builder.WriteString(fmt.Sprintf("    %s/%s:%d\n", e.Frame.Func.PkgPath, e.Frame.Func.FileName, e.Frame.Func.Source.GetLine()))
+		case e.Frame.Func != nil && e.Frame.Func.IsNative():
+			builder.WriteString(fmt.Sprintf("%s()\n", e.Frame.Func))
+			builder.WriteString(fmt.Sprintf("    %s.%s\n", e.Frame.Func.NativePkg, e.Frame.Func.NativeName))
+		case e.Frame.Func != nil:
+			builder.WriteString(fmt.Sprintf("%s\n", e.Stmt.String()))
+			builder.WriteString(fmt.Sprintf("    %s/%s:%d\n", e.Frame.Func.PkgPath, e.Frame.Func.FileName, e.Stmt.GetLine()))
+		default:
+			builder.WriteString(fmt.Sprintf("%s\n", e.Frame.GoFunc.Value.Type()))
+			builder.WriteString("    gonative\n")
+		}
+	}
+	return builder.String()
+}
+
+// Stacktrace returns the stack trace of the machine.
+// It collects the executions and frames from the machine's frames and statements.
+func (m *Machine) Stacktrace() (stacktrace Stacktrace) {
+	if len(m.Frames) == 0 {
+		return
+	}
+
+	var executions []Execution
+	nextStmtIndex := len(m.Stmts) - 1
+	for i := len(m.Frames) - 1; i >= 0; i-- {
```

Comment:
This loop is building a slice of executions for all applicable frames, even if they will end up being trimmed. Can the loop instead be broken when the limit is reached? Then it won't need to trim the slice after the loop execution.

Reply:
I need to iterate through all frames to determine the total count of callable frames. I separate the trimming process to enhance code readability as we must loop through all frames.

Reply:
Yeah, makes sense. I tried doing some testing to see how much of a difference it made when there exists a massive amount of frames. What I found was that it didn't matter because the code in the keeper is still printing the full machine state on panic: https://github.com/gnolang/gno/blob/79a71a47f987c5fad28be73b56ac7d3691774840/gno.land/pkg/sdk/vm/keeper.go#L317-L320

It is not limited to this specific place in the code; there are multiple other places in the keeper where this is happening. We can either address it in this PR or I can make a follow up PR to address this. I'm leaning towards the follow up PR to make the distinction that this PR is a feature and the follow up to this is for performance now that we have an alternative option for printing the panic message -- the stacktrace generated thanks to the work in this PR. Does that sound okay to you?

Reply:
Yes, maybe we can do another PR where we focus on putting the stacktrace where we think it's relevant. This will make the PR easier to review.